x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -252,7 +257,12 @@ public synchronized boolean maybeTriggerAsyncJob(long now) {
     protected abstract void onAbort();
 
     private void finishWithFailure(Exception exc) {
-        doSaveState(finishAndSetState(), position.get(), () -> onFailure(exc));
+        try {
+            doSaveState(finishAndSetState(), position.get(), () -> onFailure(exc));
+        } catch (Exception e) {
+            onFailure(exc);
+            onFailure(new RuntimeException("Failed to save State", e));
```

Comment:
That's the odd part of this PR. We have 2 failures, the one this method has been called with and the other one which happened during save. 

Any better idea than calling the callback twice?